Fix User Creation API - Add Error Handling and Authentication Improvements

swayam-2003 opened the pull request:
### 1. Authentication Fix
**File:** `src/app/api/admin/create-user/route.ts`

**Before:**
```typescript
// Was trying to get its own admin token
const authHeader = await getAdminToken();
```

**After:**
```typescript
// Now uses the same auth pattern as GET users API
const authHeader = req.headers.get("authorization");
```

**Why:** GET users API was working because it used the token from the frontend request. CREATE user was failing because it tried to create its own admin token.

### 2. Frontend Integration
**File:** `src/input-components/AddUserForm.tsx`

**Added:**
- Real API call to `/api/admin/create-user`
- Loading states with "Saving..." button text
- Form reset and dialog close after successful creation
- Error handling with user feedback
- Controlled dialog state management

**Before:**
```typescript
console.log("Final user:", values) // Only logged to console
```

**After:**
```typescript
const response = await axios.post("/api/admin/create-user", {
    username: values.username,
    email: values.email,
    password: values.password,
    organization: values.organization,
    group: values.group,
    firstName: values.firstName,
    lastName: values.lastName
});
```

### 3. Error Handling Improvements
**File:** `src/app/api/admin/create-user/route.ts`

**Added graceful error handling for organization/group assignment:**

```typescript
// Try to add to organization and group, but don't fail if they don't exist
try {
    await Promise.all([
        axios.post(organizationEndpoint).catch(orgError => {
            console.warn("Could not add user to organization:", orgError.response?.data);
            return null; // Don't fail the entire operation
        }),
        axios.put(groupEndpoint).catch(groupError => {
            console.warn("Could not add user to group:", groupError.response?.data);
            return null; // Don't fail the entire operation
        }),
    ]);
} catch (error) {
    console.warn("Some assignments failed, but user was created:", error);
}
```

**Why:** Organization/group assignment was causing complete failure. Now users can be created even if these assignments fail.

### 4. Data Consistency Fixes
**File:** `src/app/api/admin/users/route.ts` & `src/components/user-management/UserTable.tsx`

**Fixed attribute handling:**
- Organization and groups are now stored as arrays in Keycloak
- Table display handles both single values and arrays
- Backward compatibility maintained

**Before:**
```typescript
organization: attributes?.organization || [],
groups: attributes?.groups || [],
```

**After:**
```typescript
organization: Array.isArray(attributes?.organization) ? attributes.organization : (attributes?.organization ? [attributes.organization] : []),
groups: Array.isArray(attributes?.groups) ? attributes.groups : (attributes?.groups ? [attributes.groups] : []),
```

### 5. User Experience Improvements
**File:** `src/input-components/AddUserForm.tsx`

**Added:**
- Loading states during API calls
- Form validation
- Success/error feedback
- Automatic form reset after successful creation
- Page refresh to show new user in table

## Results

### Before:
- ❌ Users only logged to console
- ❌ 500 Internal Server Error
- ❌ No loading states
- ❌ Organization errors caused complete failure

### After:
- ✅ Users saved to Keycloak database
- ✅ Proper error handling
- ✅ Loading states and user feedback
- ✅ Graceful degradation for organization/group issues
- ✅ Users appear in user management table
- ✅ Consistent authentication with GET users API

## Testing

1. **User Creation Flow:**
   - Fill out 3-step form
   - Click "Save User"
   - Verify user appears in Keycloak
   - Verify user appears in user management table

2. **Error Handling:**
   - Test with missing required fields
   - Test with invalid organization/group IDs
   - Verify graceful error handling

3. **UI/UX:**
   - Verify loading states work
   - Verify form resets after success
   - Verify error messages display

## Files Changed
- `src/app/api/admin/create-user/route.ts` - Main API fixes
- `src/input-components/AddUserForm.tsx` - Frontend integration
- `src/app/api/admin/users/route.ts` - Data consistency
- `src/components/user-management/UserTable.tsx` - Display fixes

## Impact
This PR enables the complete user creation workflow, allowing administrators to create users through the UI that are properly stored in Keycloak and visible in the user management interface.

swayam-2003 committed:
Fix user creation API - add error handling for organization/group assignment

AsishNayak02 commented on src/app/api/admin/create-user/route.ts:
```diff
@@ -11,43 +11,78 @@ interface DecodedUserToken {
+            ADMIN_PASSWORD: process.env.ADMIN_PASSWORD ? "Set" : "Missing",
+            CLIENT_SECRET: process.env.CLIENT_SECRET ? "Set" : "Missing"
+        });
+
```
Remove the console logs before pushing

AsishNayak02 commented on src/app/api/admin/create-user/route.ts:
```diff
@@ -11,43 +11,78 @@ interface DecodedUserToken {
+        const params = new URLSearchParams();
+        params.append("client_id", process.env.CLIENT_ID!);
+        params.append("username", process.env.ADMIN_USERNAME!);
+        params.append("password", process.env.ADMIN_PASSWORD!);
```
Explore how can we create user without using this admin username and password

AsishNayak02 commented on src/app/api/admin/create-user/route.ts:
```diff
@@ -11,43 +11,78 @@ interface DecodedUserToken {
+        }
+
+        const groupID = '979482d9-31af-4977-bd66-e5ec73cc4586';
+        const organizationID = '8a6e7ea2-8880-40de-9c74-9b4fe6b43b9f';
```
Make this dynamic, there should be list of organization and group selecting one on the Ui should give their ID to assign the user to that group and organization

AsishNayak02 commented on src/app/api/admin/create-user/route.ts:
```diff
@@ -58,33 +93,53 @@ export async function POST(req: Request) {
-                    "organization": organization,
-                    "groups": group,
+                    "organization": [organization],
+                    "groups": [group],
```
why is it a array, it should be a select field which gives a single string not array

AsishNayak02 commented on src/app/api/admin/users/route.ts:
```diff
@@ -46,8 +46,8 @@ export async function GET(req: Request) {
-        organization: attributes?.organization || [],
-        groups: attributes?.groups || [],
+        organization: Array.isArray(attributes?.organization) ? attributes.organization : (attributes?.organization ? [attributes.organization] : []),
+        groups: Array.isArray(attributes?.groups) ? attributes.groups : (attributes?.groups ? [attributes.groups] : []),
```
Keep organization and groups as a strings

AsishNayak02 commented on src/input-components/AddUserForm.tsx:
```diff
@@ -41,21 +45,44 @@ export function AddUserForm() {
+                try {
+                    console.log("Creating user:", values);
+                    const response = await axios.post("/api/admin/create-user", {
+                        username: values.username,
```
make it a function and import it here

swayam-2003 committed:
Fixed search; Added EDIT and DELETE for User Mgmt; Zod integration for validation

swayam-2003 committed:
Merge dev branch with user-mgmt-changes - resolved conflicts and kept enhanced features

swayam-2003 committed:
feat: Complete Role Definition Page Implementation

- Add comprehensive role management system with CRUD operations
- Implement realm role creation with client role assignments
- Add role editing with client role management
- Create role deletion with confirmation dialog
- Add search and filtering functionality
- Implement real-time form validation with ZOD
- Add proper error handling and loading states
- Remove all icons and use emoji/text symbols
- Fix HTML nesting errors in dialogs
- Add client role context display (Account Management, Realm Management, etc.)
- Integrate with Keycloak admin API for role operations
- Add responsive design and clean UI
